lib/vdsm/supervdsm_api/udev.py

```diff
@@ -115,24 +116,32 @@ def rmAppropriateMultipathRules(thiefId):
 
 
 @expose
-def add_managed_udev_rule(vol_id, path):
-    rule_file = _UDEV_RULE_FILE_NAME_MANAGED % vol_id
-    symlink = os.path.relpath(path, '/dev/')
-    rule = 'SYMLINK=="%s", RUN+="%s %s:%s $env{DEVNAME}"\n' % (
-        symlink, EXT_CHOWN, DISKIMAGE_USER, DISKIMAGE_GROUP)
+def add_managed_udev_rule(sd_id, vol_id, path):
+    rule_file = _UDEV_RULE_FILE_NAME_MANAGED % (sd_id, vol_id)
+    device = os.path.relpath(path, '/dev/')
+    if os.path.islink(path):
+        rule = 'SYMLINK=="%s", RUN+="%s %s:%s $env{DEVNAME}"\n' % (
+            device, EXT_CHOWN, DISKIMAGE_USER, DISKIMAGE_GROUP)
+    else:
+        rule = 'KERNEL=="%s", RUN+="%s %s:%s $env{DEVNAME}"\n' % (
+            device, EXT_CHOWN, DISKIMAGE_USER, DISKIMAGE_GROUP)
     with open(rule_file, "w") as rf:
         _log.debug("Creating rule %s: %r", rule_file, rule)
         rf.write(rule)
 
 
 @expose
 def trigger_managed_udev_rule(path):
-    _udevTrigger(property_matches=(('DEVLINKS', path),))
+    # Example path: /dev/rbd/volumes/volume-d1530bb1-...-1adbcbc4eedc
+    if os.path.islink(path):
+        _udevTrigger(property_matches=(('DEVLINKS', f"*{path}*"),))
```

Comment:
Why do we need `*{path}*` now? This worked before with path as is. What is the meaning of the `*` in the match?

Reply:
usedevadm(8) says:

    udevadm trigger [options] [devpath|file|unit]

Since we trigger events for known path, can we use devpath and remove the --property-match?

I tried this:

In one shell:
```
[nsoffer@host4 ~]$ udevadm monitor
monitor will print the received events for:
UDEV - the event which udev sends out after rule processing
KERNEL - the kernel uevent

```

In another shell:
```
[root@host4 vdsm]# udevadm trigger /dev/mapper/0QEMU_QEMU_HARDDISK_data-fc-01
[root@host4 vdsm]# udevadm trigger /dev/mapper/36001405789e8ab39498483ea5858c779
```

In the monitor shell we see:
```
KERNEL[103021.013793] change   /devices/virtual/block/dm-2 (block)
UDEV  [103021.069396] change   /devices/virtual/block/dm-2 (block)
KERNEL[103213.074750] change   /devices/virtual/block/dm-15 (block)
UDEV  [103213.114084] change   /devices/virtual/block/dm-15 (block)
```

Seems to work - can you try with rbd and the iscsi server use for testing?

Reply:
I'll check using the path,
as for asterisks, explained here: https://github.com/oVirt/vdsm/pull/89#discussion_r876812403

Reply:
@nirs 
 
I can use just the path if I apply the following diff:
```diff
diff --git a/lib/vdsm/common/udevadm.py b/lib/vdsm/common/udevadm.py
index d849f0228..b5e7cee2d 100644
--- a/lib/vdsm/common/udevadm.py
+++ b/lib/vdsm/common/udevadm.py
@@ -53,7 +53,7 @@ def settle(timeout, exit_if_exists=None):
         logging.error("%s", e)


-def trigger(attr_matches=(), property_matches=(), subsystem_matches=()):
+def trigger(path=None, attr_matches=(), property_matches=(), subsystem_matches=()):
     '''
     Request device events from the kernel.

@@ -87,6 +87,7 @@ def trigger(attr_matches=(), property_matches=(), subsystem_matches=()):
     _run_command(['control', '--reload'])

     cmd = ['trigger', '--verbose', '--action', 'change']
+    cmd.append(path)

     for name, value in property_matches:
         cmd.append('--property-match={}={}'.format(name, value))
diff --git a/lib/vdsm/supervdsm_api/udev.py b/lib/vdsm/supervdsm_api/udev.py
index 506d195c3..5af298541 100644
--- a/lib/vdsm/supervdsm_api/udev.py
+++ b/lib/vdsm/supervdsm_api/udev.py
@@ -132,7 +132,7 @@ def add_managed_udev_rule(sd_id, vol_id, path):

 @expose
 def trigger_managed_udev_rule(path):
-    # Example path: /dev/rbd/volumes/volume-d1530bb1-...-1adbcbc4eedc
-    if os.path.islink(path):
-        _udevTrigger(property_matches=(('DEVLINKS', f"*{path}*"),))
-    else:
-       _udevTrigger(property_matches=(('DEVNAME', path),))
-
+    _udevTrigger(path=path)
```

The other options include the asterisks like shown above, or use the original with `os.path.realpath`
```python
    _udevTrigger(property_matches=(('DEVNAME', os.path.realpath(path)),))
```

Reply:
> @nirs
> 
> I can use just the path if I apply the following diff:
> 
> ```diff
> diff --git a/lib/vdsm/common/udevadm.py b/lib/vdsm/common/udevadm.py
> index d849f0228..b5e7cee2d 100644
> --- a/lib/vdsm/common/udevadm.py
> +++ b/lib/vdsm/common/udevadm.py
> @@ -53,7 +53,7 @@ def settle(timeout, exit_if_exists=None):
>          logging.error("%s", e)
> 
> 
> -def trigger(attr_matches=(), property_matches=(), subsystem_matches=()):
> +def trigger(path=None, attr_matches=(), property_matches=(), subsystem_matches=()):
>      '''
>      Request device events from the kernel.
> 
> @@ -87,6 +87,7 @@ def trigger(attr_matches=(), property_matches=(), subsystem_matches=()):
>      _run_command(['control', '--reload'])
> 
>      cmd = ['trigger', '--verbose', '--action', 'change']
> +    cmd.append(path)

Should be:

    if path:
        cmd.append(path)

> 
>      for name, value in property_matches:
>          cmd.append('--property-match={}={}'.format(name, value))

And it should be done *after* adding the other options.

I'm not sure if using both path and properties makes sense, but this is something callers
should handle.

> diff --git a/lib/vdsm/supervdsm_api/udev.py b/lib/vdsm/supervdsm_api/udev.py
> index 506d195c3..5af298541 100644
> --- a/lib/vdsm/supervdsm_api/udev.py
> +++ b/lib/vdsm/supervdsm_api/udev.py
> @@ -132,7 +132,7 @@ def add_managed_udev_rule(sd_id, vol_id, path):
> 
>  @expose
>  def trigger_managed_udev_rule(path):
> -    # Example path: /dev/rbd/volumes/volume-d1530bb1-...-1adbcbc4eedc
> -    if os.path.islink(path):
> -        _udevTrigger(property_matches=(('DEVLINKS', f"*{path}*"),))
> -    else:
> -       _udevTrigger(property_matches=(('DEVNAME', path),))
> -
> +    _udevTrigger(path=path)

This part looks good.

> The other options include the asterisks like shown above, or use the original with `os.path.realpath`
> 
> ```python
>     _udevTrigger(property_matches=(('DEVNAME', os.path.realpath(path)),))
> ```

So the original worked with realative path, and it is broken now becasue we started
to use the full path?